src/core/types.ts

```diff
@@ -224,11 +271,28 @@ export type GetDataState<
   TState extends DataState<TData>["dataState"],
 > = Extract<DataState<TData>, { dataState: TState }>;
 
+export type FormattedExecutionResultWithDataState<
+  TData = Record<string, unknown>,
+  TExtensions = Record<string, unknown>,
+> = Omit<FormattedExecutionResult<TData, TExtensions>, "data"> &
+  (
+    | {
+        data?: TData | null;
```

Comment:
Looking at this one more time, I don't think this is correct. If its `complete`, `data` should be a required, non-nullable `TData` type.

```suggestion
        data: TData;
```